### PR TITLE
Use named fields for casts in ast::Expression

### PR DESCRIPTION
**Pull request by seanyoung**
Signed-off-by: Sean Young <sean@mess.org>

### CODE REVIEW

**Comment by seanyoung** [comment]
I think I'd like to convert this piecemeal, otherwise it's too much work in one go.

**Review by xermicus — APPROVED** [review]
Nice, thanks! Totally fine for me to do this in smaller steps

File: src/codegen/expression.rs [hunk 1]
```diff
@@ -496,63 +496,69 @@ pub fn expression(
...
             *loc,
-            ty.clone(),
-            Box::new(expression(e, cfg, contract_no, func, ns, vartab, opt)),
+            to.clone(),
+            Box::new(expression(expr, cfg, contract_no, func, ns, vartab, opt)),
         ),
-        ast::Expression::Cast(loc, ty @ Type::Address(_), e) => {
-            if let Ok((_, address)) = eval_const_number(e, ns) {
-                Expression::NumberLiteral(*loc, ty.clone(), address)
+        ast::Expression::Cast { loc, to, expr } if matches!(to, Type::Address(_)) => {
```
Comment (LucasSte) [inline src/codegen/expression.rs:517]: I think you can still write `ast::Cast{loc, to @ Type::Address(_), expr)` here.

File: src/sema/ast.rs [hunk 2]
```diff
@@ -686,13 +686,37 @@ pub enum Expression {
...
+    },
+    Cast {
+        loc: pt::Loc,
+        to: Type,
+        expr: Box<Expression>,
+    },
+    BytesCast {
+        loc: pt::Loc,
+        to: Type,
+        from: Type,
```
Comment (LucasSte) [inline src/sema/ast.rs:717]:
```suggestion
        from: Type,
        to: Type,
```

I would switch the order here for better understanding, as we always go from a type to another one.

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: src/codegen/expression.rs (hunk 1 above)
Comment (seanyoung) [reply src/codegen/expression.rs:517]: I thought that too, but:
```
error: expected `,`
   --> src/codegen/expression.rs:517:40
    |
517 |         ast::Expression::Cast { loc, to@ Type::Address(_), expr } => {
    |         ---------------------          ^
    |         |
    |         while parsing the fields for this pattern

```
Surprising, isn't it?

**Review by LucasSte — APPROVED** [review]

**Pull request merged**